Update Code First Process for GitHub Issues

makubacki opened the pull request:
Changes to the EDK II Code First Process to accommodate GitHub issues.

---

Open for feedback.

makubacki committed:
EDK II Code First Process: Fix markdown link issues

This change fixes markdown formatting issues in preparation to clean
up the text for an upcoming content update.

https://github.com/DavidAnson/markdownlint

Signed-off-by: Michael Kubacki <michael.kubacki@microsoft.com>

makubacki committed:
EDK-II-Code-First-Process: GitHub issue update

Updates the code first process to accommodate GitHub issues.

Signed-off-by: Michael Kubacki <michael.kubacki@microsoft.com>

leiflindholm commented on EDK-II-Code-First-Process.md:
```diff
@@ -116,38 +92,41 @@ Optional Section
+    - Note: `<BranchName>` is a brief description of the change.
+    - Note: Code content must follow the coding style and naming conventions in
+      the [Source Code](#source-code) section of this document.
+3. Push the branch with the changes to [edk2-staging](https://github.com/tianocore/edk2-staging).
```
What does this step look like in practice for someone who does not have admin rights to the edk2-staging repository?

makubacki replied:
I agree that's been an ongoing issue with the process. We could allow forks to be used initially with contact information for the [EDK II Staging Maintainers](https://github.com/orgs/tianocore/teams/edk-ii-staging-maintainers) that can be used to request a forked branch be pushed to the upstream for greater visibility.

makubacki replied:
Another option would be to let the changes reside in a draft PR to edk2 that has the `type:code-first` label applied. I prefer that method to make changes more visible in the edk2 repo, against edk2 CI, with permissions control during that time left to the contributor on their fork.

I will update the content to reflect this approach for now.

makubacki replied:
Pushed that update.

leiflindholm replied:
One concern I have with that approach is that the only thing tying that PR to the tianocore/edk2 repository is then github itself.

makubacki replied:
Can you please elaborate a bit more?

spbrogan replied:
If the code-first request was accepted by edk2 then the process could be to create a longer lived branch in github.com/tianocore/edk2/code-first/<topic name> 

Would that resolve your concern?

leiflindholm replied:
@spbrogan yes, that would resolve my concern. I like that.

@makubacki My concern is because of the way github operates, with the PR branch living in the author's forked repo. This creates workflow concerns for me:
1) Needing to juggle more remotes. (Or manually cross-reference against the web UI to get the PR number, sure, but... Plus that's a github special, and I like to avoid those where we don't fundamentally depend on them.)
2) How to cope with "passing the torch" if the original PR creator ... changes roles and can't continue, but someone else takes over.

Sean's proposal resolves both.

makubacki replied:
Made the following updates based on decisions in the stewards meeting.

1. Allow the primary dev branch to either be on a fork or a branch in edk2-staging.
   - Explain the impact and outcome of that decision.
2. Add additional steps to the intended workflow after the draft PR is created.
3. Add a note that the source code tagging requirement is currently in place but subject to change in the future.

makubacki replied:
@mdkinney, @leiflindholm, @spbrogan, please review the current content to finalize the changes.

makubacki replied:
Marking this thread as resolved given the current changes.

makubacki committed:
EDK-II-Code-First-Process: Revise contributor process

Use draft pull requests instead of edk2-staging.

Signed-off-by: Michael Kubacki <michael.kubacki@microsoft.com>

makubacki commented on EDK-II-Code-First-Process.md:
```diff
@@ -110,73 +89,80 @@ Required Section
+When the change is ready for review, the PR should be marked as ready for review (taken out of draft status).
 
-# Source Code
+## Source Code
```
We've found modifying all the symbols to be very tedious. Is this something others would be open to changing? Are the changes being in a PR marked "draft" and labeled `type:code-first` sufficient?

mdkinney replied:
That is a good question.  Both approaches are functional.  The advantage of tagging the symbols is that changes that have not been published in a public doc are different and obvious so any downstream developer can detect if pre-release content is present.  If there are other ways downstream source tree without network access can tell if this type of content is present, then I would welcome suggestions for the community to discuss.

Your suggestion would be simpler and less error prone from the content producer perspective.  There are many more consumers than producers, so we have to optimize for the content consumer.

leiflindholm replied:
I agree with @mdkinney that optimising for consumers - easily detecting draft-spec code in the wild - is worh some pain. This was explicitly why we wrote it like this. But not any level of pain.

What are the tedious aspects you've found? Can we improve those while keeping the checked out tree useful for identifying draft-relevant modifications without requiring manual cross-referencing?

makubacki replied:
Do we know how many people have followed this process? Looking for ["BZ" branches](https://github.com/tianocore/edk2-staging/branches/all?query=BZ) on edk2-staging I only see three - myself, Kun, and Nickle. If that's our sample population, then at least Kun and I found it tedious to rename everything twice and differentiate actual code changes between two copies which are littered with differences because of all the symbol name changes.

Something that should be clarified in this PR's update is what happens with the draft PR (after Step 5). I was envisioning that the PR stays in draft state until the corresponding spec change(s) are published and then the PR is published from draft, reviewed, and eventually merged like any other PR.

My understanding of the need here is that a consumer needs to know the state of the code while in that draft state. And there are at least two approaches that could be taken to manage the changes in the draft state.

**Code Stays in Draft PR**

My understanding is this is the current approach taken.

- If the changes stay in the draft PR (not merged) until the spec is released, I view the draft PR with the `type:code-first` label as the status indicator to a consumer that encounters the code changes in the upstream repo.
  - With this approach, it seems to me that the value in tagging the code itself is mostly for secondary consumers that may come across the code after a primary consumer has potentially copied it elsewhere.

**Code is Merged**

- Here the code is merged to edk2/master.
- In this case, tagging is absolutely necessary as the code is accessible to a large set of developers that have no other indicator to its status.

---

To make sure we're talking about the same process, can you please confirm, we'd use the **Code Stays in Draft PR** approach, and that the tagging requirement is to help clarify the status of the code in that draft PR?

I'm not strongly opposed to tagging, I understand it has value to encode the status directly in the code changes. I just thought I'd raise the issue while we're making process changes since Kun and I found some friction in the process and to reassess if we still think that degree of tracking is needed.

mdkinney replied:
Some consumers can choose to integrate pre-release content so the code first content would be merged into a downstream branch or even a copy of the tree without use of git.  It is these use cases where the tagging is beneficial.

For code first changes that are never integrated in this way, I would agree that the tagging is extra work without any benefit.  However, we can not predict how downstream consumers will use this content.

In order to perform extensive testing of code first proposals, downstream consumers would have to integrate the code first PRs into a test build.  If something fails in testing, the tags make be useful in root causing failures.

mdkinney replied:
One comment on Code First being in a "Draft PR".  That makes sense to me so any consumer can look for PRs in the Draft state with the code-first label.  We would want to update the CI checker for stale PRs to not auto close these types of PRs.

Also, by having all the code-first PRs in the edk2 repo, we can have a cron job that attempts to take latest edk2 code and find all draft PRs with code-first label and verify that they can be merged into edk2.  That would detect code first PRs that are out of date and code first PRs that have merge conflicts with each other.  If all the code first PRs can be merged, then additional CI checks such as basic build and boot test can be performed.  This can provide feedback to the code first PR owners to keep their draft PRs stable for downstream testing.  We may need another label to indicate of a code first Draft PR is ready for integration testing.

makubacki replied:
> In order to perform extensive testing of code first proposals, downstream consumers would have to integrate the code first PRs into a test build. If something fails in testing, the tags make be useful in root causing failures.

Then we're on the same page for why it is being done and if we think that value is worth it (optimizing for the consumer), we can acknowledge that and move on.

> We would want to update the CI checker for stale PRs to not auto close these types of PRs.

Yes. I will include that in the edk2 PR.

makubacki replied:
Pushed in this commit: https://github.com/tianocore/edk2/pull/10815/commits/7adbcf53b3c4d8853b1871c3cf53d2c1aad9f2ee

makubacki replied:
Resolving this thread since we're going to leave the tagging convention in place for now.

makubacki committed:
EDK-II-Code-First-Process: Remove spec draft template

The template will be maintained in a single location - in the GitHub
issue form where it is used.

Signed-off-by: Michael Kubacki <michael.kubacki@microsoft.com>

makubacki committed:
EDK-II-Code-First-Process: Revise branch and PR details

1. Allow the primary dev branch to either be on a fork or a branch in
   edk2-staging.
   - Explain the impact and outcome of that decision.
2. Add additional steps to the intended workflow after the draft PR
   is created.
3. Add a note that the source code tagging requirement is currently
   in place but subject to change in the future.

Signed-off-by: Michael Kubacki <michael.kubacki@microsoft.com>

makubacki commented on EDK-II-Code-First-Process.md:
```diff
@@ -12,171 +14,126 @@ specification update is then submitted to the appropriate working group as an
+        your fork branch as needed for others to collaborate there.
+      - If you use an `edk2-staging` branch, you might need to reach out to an edk2-staging maintainer so they can
+        grant permissions to the users that need to push changes there.
+        - If you do not have write permission, start an [edk2-staging discussion](https://github.com/tianocore/edk2-staging/discussions)
+          to introduce your branch and identify the GitHub users that need permission to contribute to the branch.
```
Let me know if there is a better way for access to be requested.

mdkinney replied:
Since the request has to get to an admin, the best place to make a request for permissions in edk2-staging is to send email to the edk2 devel mailing list.

leiflindholm replied:
Alternatively, we can handle that as part of the Issue?

mdkinney replied:
Listing the collaborators in the issue is a great idea so anyone looking at it can know who is involved to ask questions.

Putting requests to Tianocore Admins in the issue discussion to add edk2-staging collaborators should work.

makubacki replied:
Updated to this process.

mdkinney commented on EDK-II-Code-First-Process.md:
```diff
@@ -12,171 +14,126 @@ specification update is then submitted to the appropriate working group as an
+6. Continue to develop the change in the "code first dev branch" until it is ready for review. Changes pushed to the
+   branch will automatically update the PR.
+7. When the change is ready for review, mark the PR as ready for review (taken out of draft status).
+8. Reviewers will review the PR and provide feedback.
```
Step 7 here has a dependency on the spec content being public.  The state should not be changed to Ready For Review before that point.

makubacki replied:
Explicitly mentioned that in the latest push.

leiflindholm commented:
@ajfish @ardbiesheuvel

leiflindholm commented on EDK-II-Code-First-Process.md:
```diff
@@ -97,22 +96,26 @@ Optional Section
    - Note: The primary repository will most frequently be [edk2](https://github.com/tianocore/edk2).
    - Note: Ensure all specifications impacted by the change are selected in the form.
-     - Note: A specification draft change must be included for each specification impacted.
+     - Note: A specification draft change must be included in the issue for each specification impacted.
```
In the issue or the draft PR?

makubacki replied:
Draft PR. It was left over from an earlier revision. Reworked that section for accuracy and clarity.

makubacki committed:
EDK-II-Code-First-Process: Update edk2-staging permission request

- Correct specification template location to be the dev branch (PR).
  - Provide additional details on the file.
- Update the process followed to request permissions to collaborators
  on an edk2-staging branch.
- Explicitly state specs being published publicly as a prerequisite
  for submitting code changes based on the spec.

Signed-off-by: Michael Kubacki <michael.kubacki@microsoft.com>

makubacki commented:
@mdkinney & @leiflindholm, thanks for the approvals. It looks like all active code first PRs are approved with no opens. Any final requests? Any concerns with merging everything on Friday?

mdkinney commented:
No concerns.

makubacki commented:
Closing the PR. Changes are merged to the wiki page.
